// @nivo/geo/dist/nivo-geo.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var core = require('@nivo/core');
var PropTypes = _interopDefault(require('prop-types'));
var _get = _interopDefault(require('lodash/get'));
var _isFunction = _interopDefault(require('lodash/isFunction'));
var d3Format = require('d3-format');
var d3Geo = require('d3-geo');

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
var projectionById = {
  azimuthalEqualArea: d3Geo.geoAzimuthalEqualArea,
  azimuthalEquidistant: d3Geo.geoAzimuthalEquidistant,
  gnomonic: d3Geo.geoGnomonic,
  orthographic: d3Geo.geoOrthographic,
  stereographic: d3Geo.geoStereographic,
  equalEarth: d3Geo.geoEqualEarth,
  equirectangular: d3Geo.geoEquirectangular,
  mercator: d3Geo.geoMercator,
  transverseMercator: d3Geo.geoTransverseMercator,
  naturalEarth1: d3Geo.geoNaturalEarth1
};
var useGeoMap = function useGeoMap(_ref) {
  var width = _ref.width,
      height = _ref.height,
      projectionType = _ref.projectionType,
      projectionScale = _ref.projectionScale,
      projectionTranslation = _ref.projectionTranslation,
      projectionRotation = _ref.projectionRotation,
      fillColor = _ref.fillColor,
      borderWidth = _ref.borderWidth,
      borderColor = _ref.borderColor;
  var projection = React.useMemo(function () {
    return projectionById[projectionType]().scale(projectionScale).translate([width * projectionTranslation[0], height * projectionTranslation[1]]).rotate(projectionRotation);
  }, [width, height, projectionType, projectionScale, projectionTranslation[0], projectionTranslation[1], projectionRotation[0], projectionRotation[1], projectionRotation[2]]);
  var path = React.useMemo(function () {
    return d3Geo.geoPath(projection);
  }, [projection]);
  var graticule = React.useMemo(function () {
    return d3Geo.geoGraticule();
  });
  var getBorderWidth = React.useMemo(function () {
    return typeof borderWidth === 'function' ? borderWidth : function () {
      return borderWidth;
    };
  }, [borderWidth]);
  var getBorderColor = React.useMemo(function () {
    return typeof borderColor === 'function' ? borderColor : function () {
      return borderColor;
    };
  }, [borderColor]);
  var getFillColor = React.useMemo(function () {
    return typeof fillColor === 'function' ? fillColor : function () {
      return fillColor;
    };
  }, [fillColor]);
  return {
    projection: projection,
    path: path,
    graticule: graticule,
    getBorderWidth: getBorderWidth,
    getBorderColor: getBorderColor,
    getFillColor: getFillColor
  };
};
var useChoropleth = function useChoropleth(_ref2) {
  var features = _ref2.features,
      data = _ref2.data,
      match = _ref2.match,
      label = _ref2.label,
      value = _ref2.value,
      valueFormat = _ref2.valueFormat,
      colors = _ref2.colors,
      unknownColor = _ref2.unknownColor;
  var findMatchingDatum = React.useMemo(function () {
    if (_isFunction(match)) return match;
    return function (feature, datum) {
      var featureKey = _get(feature, match);
      var datumKey = _get(datum, match);
      return featureKey && featureKey === datumKey;
    };
  }, [match]);
  var getLabel = React.useMemo(function () {
    return _isFunction(label) ? label : function (datum) {
      return _get(datum, label);
    };
  }, [label]);
  var getValue = React.useMemo(function () {
    return _isFunction(value) ? value : function (datum) {
      return _get(datum, value);
    };
  }, [value]);
  var valueFormatter = React.useMemo(function () {
    if (valueFormat === undefined) return function (d) {
      return d;
    };
    if (_isFunction(valueFormat)) return valueFormat;
    return d3Format.format(valueFormat);
  }, [valueFormat]);
  var getFillColor = React.useMemo(function () {
    var colorScale = core.guessQuantizeColorScale(colors).domain([0, 1000000]);
    return function (feature) {
      if (feature.value === undefined) return unknownColor;
      return colorScale(feature.value);
    };
  }, [colors, unknownColor]);
  var boundFeatures = React.useMemo(function () {
    return features.map(function (feature) {
      var datum = data.find(function (datum) {
        return findMatchingDatum(feature, datum);
      });
      var datumValue = getValue(datum);
      if (datum) {
        var featureWithData = _objectSpread({}, feature, {
          data: datum,
          value: datumValue,
          formattedValue: valueFormatter(datumValue)
        });
        featureWithData.color = getFillColor(featureWithData);
        featureWithData.label = getLabel(featureWithData);
        return featureWithData;
      }
      return feature;
    });
  }, [features, data, findMatchingDatum, getValue, valueFormatter, getFillColor]);
  return {
    getFillColor: getFillColor,
    boundFeatures: boundFeatures
  };
};

var ChoroplethTooltip = React.memo(function (_ref) {
  var feature = _ref.feature;
  if (feature.data === undefined) return null;
  return React__default.createElement(core.BasicTooltip, {
    id: feature.label,
    color: feature.color,
    enableChip: true,
    value: feature.formattedValue
  });
});
ChoroplethTooltip.propTypes = {
  feature: PropTypes.object.isRequired
};
ChoroplethTooltip.displayName = 'ChoroplethTooltip';

function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty$1(target, key, source[key]); }); } return target; }
function _defineProperty$1(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
var commonPropTypes = {
  features: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    type: PropTypes.oneOf(['Feature']).isRequired,
    properties: PropTypes.object,
    geometry: PropTypes.object.isRequired
  })).isRequired,
  projectionType: PropTypes.oneOf(Object.keys(projectionById)).isRequired,
  projectionScale: PropTypes.number.isRequired,
  projectionTranslation: PropTypes.arrayOf(PropTypes.number).isRequired,
  projectionRotation: PropTypes.arrayOf(PropTypes.number).isRequired,
  fillColor: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  borderWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.func]).isRequired,
  borderColor: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  enableGraticule: PropTypes.bool.isRequired,
  graticuleLineWidth: PropTypes.number.isRequired,
  graticuleLineColor: PropTypes.string.isRequired,
  isInteractive: PropTypes.bool.isRequired,
  onMouseEnter: PropTypes.func.isRequired,
  onMouseMove: PropTypes.func.isRequired,
  onMouseLeave: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
  tooltip: PropTypes.any,
  layers: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.oneOf(['graticule', 'features']), PropTypes.func])).isRequired
};
var GeoMapPropTypes = _objectSpread$1({}, commonPropTypes);
var GeoMapCanvasPropTypes = _objectSpread$1({
  pixelRatio: PropTypes.number.isRequired
}, commonPropTypes);
var commonChoroplethPropTypes = {
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  match: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  valueFormat: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  colors: core.quantizeColorScalePropType.isRequired,
  unknownColor: PropTypes.string.isRequired
};
var ChoroplethPropTypes = _objectSpread$1({}, GeoMapPropTypes, commonChoroplethPropTypes);
var ChoroplethCanvasPropTypes = _objectSpread$1({}, GeoMapCanvasPropTypes, commonChoroplethPropTypes);
var commonDefaultProps = {
  projectionType: 'mercator',
  projectionScale: 100,
  projectionTranslation: [0.5, 0.5],
  projectionRotation: [0, 0, 0],
  enableGraticule: false,
  graticuleLineWidth: 0.5,
  graticuleLineColor: '#999999',
  fillColor: '#dddddd',
  borderWidth: 0,
  borderColor: '#000000',
  isInteractive: true,
  onMouseEnter: function onMouseEnter() {},
  onMouseLeave: function onMouseLeave() {},
  onMouseMove: function onMouseMove() {},
  onClick: function onClick() {},
  layers: ['graticule', 'features']
};
var GeoMapDefaultProps = _objectSpread$1({}, commonDefaultProps);
var GeoMapCanvasDefaultProps = _objectSpread$1({}, commonDefaultProps, {
  pixelRatio: global.window && global.window.devicePixelRatio ? global.window.devicePixelRatio : 1
});
var commonChoroplethDefaultProps = {
  match: 'id',
  label: 'id',
  value: 'value',
  colors: 'PuBuGn',
  unknownColor: '#999',
  tooltip: ChoroplethTooltip
};
var ChoroplethDefaultProps = _objectSpread$1({}, GeoMapDefaultProps, commonChoroplethDefaultProps);
var ChoroplethCanvasDefaultProps = _objectSpread$1({}, GeoMapCanvasDefaultProps, commonChoroplethDefaultProps);

var GeoGraticule = React.memo(function (_ref) {
  var path = _ref.path,
      graticule = _ref.graticule,
      lineWidth = _ref.lineWidth,
      lineColor = _ref.lineColor;
  return React__default.createElement("path", {
    fill: "none",
    strokeWidth: lineWidth,
    stroke: lineColor,
    d: path(graticule())
  });
});
GeoGraticule.propTypes = {
  path: PropTypes.func.isRequired,
  graticule: PropTypes.func.isRequired,
  lineWidth: PropTypes.number.isRequired,
  lineColor: PropTypes.string.isRequired
};
GeoGraticule.displayName = 'GeoGraticule';

var GeoMapFeature = React.memo(function (_ref) {
  var feature = _ref.feature,
      path = _ref.path,
      fillColor = _ref.fillColor,
      borderWidth = _ref.borderWidth,
      borderColor = _ref.borderColor,
      _onClick = _ref.onClick,
      _onMouseEnter = _ref.onMouseEnter,
      _onMouseMove = _ref.onMouseMove,
      _onMouseLeave = _ref.onMouseLeave;
  return React__default.createElement("path", {
    key: feature.id,
    fill: fillColor,
    strokeWidth: borderWidth,
    stroke: borderColor,
    strokeLinejoin: "bevel",
    d: path(feature),
    onMouseEnter: function onMouseEnter(event) {
      return _onMouseEnter(feature, event);
    },
    onMouseMove: function onMouseMove(event) {
      return _onMouseMove(feature, event);
    },
    onMouseLeave: function onMouseLeave(event) {
      return _onMouseLeave(feature, event);
    },
    onClick: function onClick(event) {
      return _onClick(feature, event);
    }
  });
});
GeoMapFeature.propTypes = {
  feature: PropTypes.shape({
    id: PropTypes.string.isRequired,
    type: PropTypes.oneOf(['Feature']).isRequired,
    properties: PropTypes.object,
    geometry: PropTypes.object.isRequired
  }).isRequired,
  path: PropTypes.func.isRequired,
  fillColor: PropTypes.string.isRequired,
  borderWidth: PropTypes.number.isRequired,
  borderColor: PropTypes.string.isRequired,
  onMouseEnter: PropTypes.func.isRequired,
  onMouseMove: PropTypes.func.isRequired,
  onMouseLeave: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired
};
GeoMapFeature.displayName = 'GeoMapFeature';

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }
function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
var GeoMap = React.memo(function (props) {
  var width = props.width,
      height = props.height,
      partialMargin = props.margin,
      features = props.features,
      layers = props.layers,
      projectionType = props.projectionType,
      projectionScale = props.projectionScale,
      projectionTranslation = props.projectionTranslation,
      projectionRotation = props.projectionRotation,
      fillColor = props.fillColor,
      borderWidth = props.borderWidth,
      borderColor = props.borderColor,
      enableGraticule = props.enableGraticule,
      graticuleLineWidth = props.graticuleLineWidth,
      graticuleLineColor = props.graticuleLineColor,
      isInteractive = props.isInteractive,
      onClick = props.onClick,
      Tooltip = props.tooltip;
  var _useDimensions = core.useDimensions(width, height, partialMargin),
      margin = _useDimensions.margin,
      outerWidth = _useDimensions.outerWidth,
      outerHeight = _useDimensions.outerHeight;
  var _useGeoMap = useGeoMap({
    width: width,
    height: height,
    projectionType: projectionType,
    projectionScale: projectionScale,
    projectionTranslation: projectionTranslation,
    projectionRotation: projectionRotation,
    fillColor: fillColor,
    borderWidth: borderWidth,
    borderColor: borderColor
  }),
      graticule = _useGeoMap.graticule,
      path = _useGeoMap.path,
      getFillColor = _useGeoMap.getFillColor,
      getBorderWidth = _useGeoMap.getBorderWidth,
      getBorderColor = _useGeoMap.getBorderColor;
  var theme = core.useTheme();
  var _useTooltip = core.useTooltip(),
      _useTooltip2 = _slicedToArray(_useTooltip, 2),
      showTooltip = _useTooltip2[0],
      hideTooltip = _useTooltip2[1];
  var handleClick = React.useCallback(function (feature, event) {
    return isInteractive && onClick && onClick(feature, event);
  }, [isInteractive, onClick]);
  var handleMouseEnter = React.useCallback(function (feature, event) {
    return isInteractive && Tooltip && showTooltip(React__default.createElement(Tooltip, {
      feature: feature
    }), event);
  }, [isInteractive, showTooltip, Tooltip]);
  var handleMouseMove = React.useCallback(function (feature, event) {
    return isInteractive && Tooltip && showTooltip(React__default.createElement(Tooltip, {
      feature: feature
    }), event);
  }, [isInteractive, showTooltip, Tooltip]);
  var handleMouseLeave = React.useCallback(function () {
    return isInteractive && hideTooltip();
  }, [isInteractive, hideTooltip]);
  return React__default.createElement(core.SvgWrapper, {
    width: outerWidth,
    height: outerHeight,
    margin: margin,
    theme: theme
  }, layers.map(function (layer, i) {
    if (layer === 'graticule') {
      if (enableGraticule !== true) return null;
      return React__default.createElement(GeoGraticule, {
        key: "graticule",
        path: path,
        graticule: graticule,
        lineWidth: graticuleLineWidth,
        lineColor: graticuleLineColor
      });
    }
    if (layer === 'features') {
      return React__default.createElement(React.Fragment, {
        key: "features"
      }, features.map(function (feature) {
        return React__default.createElement(GeoMapFeature, {
          key: feature.id,
          feature: feature,
          path: path,
          fillColor: getFillColor(feature),
          borderWidth: getBorderWidth(feature),
          borderColor: getBorderColor(feature),
          onMouseEnter: handleMouseEnter,
          onMouseMove: handleMouseMove,
          onMouseLeave: handleMouseLeave,
          onClick: handleClick
        });
      }));
    }
    return React__default.createElement(React.Fragment, {
      key: i
    }, "layer(props)");
  }));
});
GeoMap.propTypes = GeoMapPropTypes;
GeoMap.defaultProps = GeoMapDefaultProps;
var GeoMap$1 = core.withContainer(GeoMap);

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
var ResponsiveGeoMap = function ResponsiveGeoMap(props) {
  return React__default.createElement(core.ResponsiveWrapper, null, function (_ref) {
    var width = _ref.width,
        height = _ref.height;
    return React__default.createElement(GeoMap$1, _extends({
      width: width,
      height: height
    }, props));
  });
};

function _slicedToArray$1(arr, i) { return _arrayWithHoles$1(arr) || _iterableToArrayLimit$1(arr, i) || _nonIterableRest$1(); }
function _nonIterableRest$1() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }
function _iterableToArrayLimit$1(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles$1(arr) { if (Array.isArray(arr)) return arr; }
var getFeatureFromMouseEvent = function getFeatureFromMouseEvent(event, el, features, projection) {
  var _getRelativeCursor = core.getRelativeCursor(el, event),
      _getRelativeCursor2 = _slicedToArray$1(_getRelativeCursor, 2),
      x = _getRelativeCursor2[0],
      y = _getRelativeCursor2[1];
  return features.find(function (f) {
    return d3Geo.geoContains(f, projection.invert([x, y]));
  });
};
var GeoMapCanvas = function GeoMapCanvas(props) {
  var width = props.width,
      height = props.height,
      partialMargin = props.margin,
      pixelRatio = props.pixelRatio,
      features = props.features,
      layers = props.layers,
      projectionType = props.projectionType,
      projectionScale = props.projectionScale,
      projectionTranslation = props.projectionTranslation,
      projectionRotation = props.projectionRotation,
      fillColor = props.fillColor,
      borderWidth = props.borderWidth,
      borderColor = props.borderColor,
      enableGraticule = props.enableGraticule,
      graticuleLineWidth = props.graticuleLineWidth,
      graticuleLineColor = props.graticuleLineColor,
      isInteractive = props.isInteractive,
      onClick = props.onClick,
      onMouseMove = props.onMouseMove,
      Tooltip = props.tooltip;
  var canvasEl = React.useRef(null);
  var theme = core.useTheme();
  var _useDimensions = core.useDimensions(width, height, partialMargin),
      margin = _useDimensions.margin,
      outerWidth = _useDimensions.outerWidth,
      outerHeight = _useDimensions.outerHeight;
  var _useGeoMap = useGeoMap({
    width: width,
    height: height,
    projectionType: projectionType,
    projectionScale: projectionScale,
    projectionTranslation: projectionTranslation,
    projectionRotation: projectionRotation,
    fillColor: fillColor,
    borderWidth: borderWidth,
    borderColor: borderColor
  }),
      projection = _useGeoMap.projection,
      graticule = _useGeoMap.graticule,
      path = _useGeoMap.path,
      getFillColor = _useGeoMap.getFillColor,
      getBorderWidth = _useGeoMap.getBorderWidth,
      getBorderColor = _useGeoMap.getBorderColor;
  React.useEffect(function () {
    if (!canvasEl) return;
    canvasEl.current.width = outerWidth * pixelRatio;
    canvasEl.current.height = outerHeight * pixelRatio;
    var ctx = canvasEl.current.getContext('2d');
    ctx.scale(pixelRatio, pixelRatio);
    ctx.fillStyle = theme.background;
    ctx.fillRect(0, 0, outerWidth, outerHeight);
    ctx.translate(margin.left, margin.top);
    path.context(ctx);
    layers.forEach(function (layer) {
      if (layer === 'graticule') {
        if (enableGraticule === true) {
          ctx.lineWidth = graticuleLineWidth;
          ctx.strokeStyle = graticuleLineColor;
          ctx.beginPath();
          path(graticule());
          ctx.stroke();
        }
      } else if (layer === 'features') {
        features.forEach(function (feature) {
          ctx.beginPath();
          path(feature);
          ctx.fillStyle = getFillColor(feature);
          ctx.fill();
          var borderWidth = getBorderWidth(feature);
          if (borderWidth > 0) {
            ctx.strokeStyle = getBorderColor(feature);
            ctx.lineWidth = borderWidth;
            ctx.stroke();
          }
        });
      } else {
        layer(ctx, props);
      }
    });
  }, [canvasEl, outerWidth, outerHeight, margin, pixelRatio, theme, path, graticule, getFillColor, getBorderWidth, getBorderColor]);
  var _useTooltip = core.useTooltip(),
      _useTooltip2 = _slicedToArray$1(_useTooltip, 2),
      showTooltip = _useTooltip2[0],
      hideTooltip = _useTooltip2[1];
  var handleMouseMove = React.useCallback(function () {
    if (!isInteractive || !Tooltip) return;
    var feature = getFeatureFromMouseEvent(event, canvasEl.current, features, projection);
    if (feature) {
      showTooltip(React__default.createElement(Tooltip, {
        feature: feature
      }), event);
    } else {
      hideTooltip();
    }
    onMouseMove && onMouseMove(feature || null, event);
  }, [showTooltip, hideTooltip, isInteractive, Tooltip, canvasEl, features, projection]);
  var handleMouseLeave = React.useCallback(function () {
    return isInteractive && hideTooltip();
  }, [isInteractive, hideTooltip]);
  var handleClick = React.useCallback(function () {
    if (!isInteractive || !onClick) return;
    var feature = getFeatureFromMouseEvent(event, canvasEl.current, features, projection);
    if (feature) {
      onClick(feature, event);
    }
  }, [isInteractive, canvasEl, features, projection, onClick]);
  return React__default.createElement("canvas", {
    ref: canvasEl,
    width: outerWidth * pixelRatio,
    height: outerHeight * pixelRatio,
    style: {
      width: outerWidth,
      height: outerHeight,
      cursor: isInteractive ? 'auto' : 'normal'
    },
    onMouseMove: handleMouseMove,
    onMouseLeave: handleMouseLeave,
    onClick: handleClick
  });
};
GeoMapCanvas.propTypes = GeoMapCanvasPropTypes;
GeoMapCanvas.defaultProps = GeoMapCanvasDefaultProps;
var GeoMapCanvas$1 = core.withContainer(GeoMapCanvas);

function _extends$1() { _extends$1 = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends$1.apply(this, arguments); }
var ResponsiveGeoMapCanvas = function ResponsiveGeoMapCanvas(props) {
  return React__default.createElement(core.ResponsiveWrapper, null, function (_ref) {
    var width = _ref.width,
        height = _ref.height;
    return React__default.createElement(GeoMapCanvas$1, _extends$1({
      width: width,
      height: height
    }, props));
  });
};

function _extends$2() { _extends$2 = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends$2.apply(this, arguments); }
var Choropleth = function Choropleth(props) {
  var _useChoropleth = useChoropleth(props),
      getFillColor = _useChoropleth.getFillColor,
      boundFeatures = _useChoropleth.boundFeatures;
  return React__default.createElement(GeoMap$1, _extends$2({}, props, {
    features: boundFeatures,
    fillColor: getFillColor
  }));
};
Choropleth.propTypes = ChoroplethPropTypes;
Choropleth.defaultProps = ChoroplethDefaultProps;

function _extends$3() { _extends$3 = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends$3.apply(this, arguments); }
var ResponsiveChoropleth = function ResponsiveChoropleth(props) {
  return React__default.createElement(core.ResponsiveWrapper, null, function (_ref) {
    var width = _ref.width,
        height = _ref.height;
    return React__default.createElement(Choropleth, _extends$3({
      width: width,
      height: height
    }, props));
  });
};

function _extends$4() { _extends$4 = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends$4.apply(this, arguments); }
var ChoroplethCanvas = function ChoroplethCanvas(props) {
  var _useChoropleth = useChoropleth(props),
      getFillColor = _useChoropleth.getFillColor,
      boundFeatures = _useChoropleth.boundFeatures;
  return React__default.createElement(GeoMapCanvas$1, _extends$4({}, props, {
    features: boundFeatures,
    fillColor: getFillColor
  }));
};
ChoroplethCanvas.propTypes = ChoroplethCanvasPropTypes;
ChoroplethCanvas.defaultProps = ChoroplethCanvasDefaultProps;

function _extends$5() { _extends$5 = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends$5.apply(this, arguments); }
var ResponsiveChoroplethCanvas = function ResponsiveChoroplethCanvas(props) {
  return React__default.createElement(core.ResponsiveWrapper, null, function (_ref) {
    var width = _ref.width,
        height = _ref.height;
    return React__default.createElement(ChoroplethCanvas, _extends$5({
      width: width,
      height: height
    }, props));
  });
};

exports.Choropleth = Choropleth;
exports.ChoroplethCanvas = ChoroplethCanvas;
exports.ChoroplethCanvasDefaultProps = ChoroplethCanvasDefaultProps;
exports.ChoroplethCanvasPropTypes = ChoroplethCanvasPropTypes;
exports.ChoroplethDefaultProps = ChoroplethDefaultProps;
exports.ChoroplethPropTypes = ChoroplethPropTypes;
exports.GeoMap = GeoMap$1;
exports.GeoMapCanvas = GeoMapCanvas$1;
exports.GeoMapCanvasDefaultProps = GeoMapCanvasDefaultProps;
exports.GeoMapCanvasPropTypes = GeoMapCanvasPropTypes;
exports.GeoMapDefaultProps = GeoMapDefaultProps;
exports.GeoMapPropTypes = GeoMapPropTypes;
exports.ResponsiveChoropleth = ResponsiveChoropleth;
exports.ResponsiveChoroplethCanvas = ResponsiveChoroplethCanvas;
exports.ResponsiveGeoMap = ResponsiveGeoMap;
exports.ResponsiveGeoMapCanvas = ResponsiveGeoMapCanvas;
exports.projectionById = projectionById;
exports.useChoropleth = useChoropleth;
exports.useGeoMap = useGeoMap;
